crates/libs/implement/src/gen.rs

```diff
@@ -0,0 +1,579 @@
+                        ::core::ptr::NonNull::new_unchecked(ptr)
+                    )
+                }
+            }
+        }
+    }
+}
+
+/// Generates the `query_interface` method.
+fn gen_query_interface(inputs: &ImplementInputs) -> syn::ImplItemFn {
+    let queries = inputs.interface_chains.iter().map(|interface_chain| {
+        let chain_ty = interface_chain.implement.to_vtbl_ident();
+        let chain_field = &interface_chain.field_ident;
+        quote_spanned! {
+            interface_chain.implement.span =>
+            if #chain_ty::matches(&iid) {
+                break 'found &self.#chain_field as *const _ as *const ::core::ffi::c_void;
+            }
+        }
+    });
+
+    // Dynamic casting requires that the object not contain non-static lifetimes.
+    let enable_dyn_casting = inputs.original_type.generics.lifetimes().count() == 0;
+    let dynamic_cast_query = if enable_dyn_casting {
+        quote! {
+            if iid == ::windows_core::DYNAMIC_CAST_IID {
+                // DYNAMIC_CAST_IID is special. We _do not_ increase the reference count for this pseudo-interface.
+                // Also, instead of returning an interface pointer, we simply write the `&dyn Any` directly to the
+                // 'interface' pointer. Since the size of `&dyn Any` is 2 pointers, not one, the caller must be
+                // prepared for this. This is not a normal QueryInterface call.
+                //
+                // See the `Interface::cast_to_any` method, which is the only caller that should use DYNAMIC_CAST_ID.
+                (interface as *mut *const dyn core::any::Any).write(self as &dyn ::core::any::Any as *const dyn ::core::any::Any);
+                return ::windows_core::HRESULT(0);
+            }
+        }
+    } else {
+        quote!()
+    };
+
+    let identity_query = quote! {
+        if iid == <::windows_core::IUnknown as ::windows_core::Interface>::IID
+        || iid == <::windows_core::IInspectable as ::windows_core::Interface>::IID
+        || iid == <::windows_core::imp::IAgileObject as ::windows_core::Interface>::IID {
+            break 'found &self.identity as *const _ as *const ::core::ffi::c_void;
+        }
+    };
+
+    let tear_off_query = quote! {
+        let tear_off_ptr = self.count.query(&iid, &self.identity as *const _ as *mut _);
+        if !tear_off_ptr.is_null() {
+            *interface = tear_off_ptr;
+            return ::windows_core::HRESULT(0);
+        }
+    };
+
+    parse_quote! {
+        unsafe fn QueryInterface(
+            &self,
+            iid: *const ::windows_core::GUID,
+            interface: *mut *mut ::core::ffi::c_void,
+        ) -> ::windows_core::HRESULT {
+            unsafe {
+                if iid.is_null() || interface.is_null() {
+                    return ::windows_core::imp::E_POINTER;
+                }
+
+                let iid = *iid;
+
+                let interface_ptr: *const ::core::ffi::c_void = 'found: {
```

Comment:
Hmm are using loop labels the simplest way to express this? Just curious what motivated this.

Reply:
I actually don't see a loop here - may have to look at the generated code to internalize this. 😊

Reply:
It eliminates another "is this null?" check.  You can do this purely with control-flow.

Reply:
There is no loop.  Rust has a concept of "named scopes you can break out of".  They're kind of like gotos, but far more constrained (and structured).

Reply:
Yeah I started having cold sweats when I thought about gotos. 🥲